src/Microsoft.AspNetCore.Razor.Tools/GenerateCommand.cs

```diff
@@ -0,0 +1,204 @@
+            });
+
+            var templateEngine = new MvcRazorTemplateEngine(engine, RazorProject.Create(projectDirectory));
+
+            var sourceItems = GetRazorFiles(projectDirectory, sources);
+            var results = GenerateCode(templateEngine, sourceItems);
+
+            var success = true;
+
+            foreach (var result in results)
+            {
+                if (result.CSharpDocument.Diagnostics.Count > 0)
+                {
+                    success = false;
+                    foreach (var error in result.CSharpDocument.Diagnostics)
+                    {
+                        Console.Error.WriteLine(error.ToString());
+                    }
+                }
+
+                var outputFilePath = Path.Combine(outputDirectory, Path.ChangeExtension(result.ViewFileInfo.ViewEnginePath.Substring(1), ".cs"));
+                File.WriteAllText(outputFilePath, result.CSharpDocument.GeneratedCode);
+            }
+
+            return success ? 0 : -1;
+        }
+
+        private IReadOnlyList<TagHelperDescriptor> GetTagHelpers(string tagHelperManifest)
+        {
+            if (!File.Exists(tagHelperManifest))
+            {
+                return Array.Empty<TagHelperDescriptor>();
+            }
+
+            using (var stream = File.OpenRead(tagHelperManifest))
+            {
+                var reader = new JsonTextReader(new StreamReader(stream));
+
+                var serializer = new JsonSerializer();
+                serializer.Converters.Add(new RazorDiagnosticJsonConverter());
+                serializer.Converters.Add(new TagHelperDescriptorJsonConverter());
+
+                return serializer.Deserialize<IReadOnlyList<TagHelperDescriptor>>(reader);
+            }
+        }
+
+        private List<SourceItem> GetRazorFiles(string projectDirectory, string[] sources)
+        {
+            var trimLength = projectDirectory.EndsWith("/") ? projectDirectory.Length - 1 : projectDirectory.Length;
+
+            var items = new List<SourceItem>(sources.Length);
+            for (var i = 0; i < sources.Length; i++)
+            {
+                var fullPath = Path.Combine(projectDirectory, sources[i]);
+                if (fullPath.StartsWith(projectDirectory, StringComparison.OrdinalIgnoreCase))
+                {
+                    var viewEnginePath = fullPath.Substring(trimLength).Replace('\\', '/');
+                    items.Add(new SourceItem(fullPath, viewEnginePath));
+                }
+            }
+
+            return items;
+        }
+
+        private OutputItem[] GenerateCode(RazorTemplateEngine templateEngine, IReadOnlyList<SourceItem> sources)
+        {
+            var outputs = new OutputItem[sources.Count];
+            Parallel.For(0, outputs.Length, new ParallelOptions() { MaxDegreeOfParallelism = 4 }, i =>
+            {
+                var source = sources[i];
+
+                var csharpDocument = templateEngine.GenerateCode(source.ViewEnginePath);
```

Comment:
If only this engine returned a code document you wouldn't need an output item at all. I'll make that possible with the project engine 😄